sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -1628,6 +1631,94 @@ def write_pipeline_spec_to_file(pipeline_spec: pipeline_spec_pb2.PipelineSpec,
         with open(package_path, 'w') as yaml_file:
             yaml.dump(json_dict, yaml_file, sort_keys=True)
 
+        with open(package_path, 'r+') as f:
+            old = f.read()
+            f.seek(0)
+            f.write(yaml_comments + old)
+
```

Comment:
Because you have extracted the comment from the `json_dict` (instead of by parsing it from the already written file), you can add the header comment in the original `with open` context before `yaml.dump`ing.

This refactoring would allow the code to write the file in the same order as its contents (comments, then YAML) in fewer steps.

Reply:
Oh! Thanks for that